### PR TITLE
fix(ci): enforce pytest exit status in test-python; make fast suite pass torch-less

### LINKED ISSUE #484: CI: test-python job is false-green — pytest collection aborts (18 errors), tee swallows the exit code, zero tests run
# Problem

The `test-python` job in `.github/workflows/ci.yml` is **false-green**: it reports success without running any tests.

Evidence from run 28740450491 (main, 2026-07-05, "success"):

```
collecting ... collected 878 items / 18 errors / 36 deselected / 5 skipped / 842 selected
==================================== ERRORS ====================================
_____________ ERROR collecting tests/test_bc_fit_class_balanced.py _____________
ImportError while importing test module ...
```

- pytest aborts at collection with **18 import errors** (torch-dependent test modules; the job installs only `.[dev]`, and the session header confirms `torch (RL): Not installed`), executes **zero tests**, and exits nonzero.
- The workflow step pipes through `| tee pytest-output.txt`. The step shell is `bash -e` **without `pipefail`**, so the step's exit status is `tee`'s — always 0. Job passes in ~10 seconds.
- The `--cov-fail-under=75` gate is also moot (coverage never runs).

Consequences:
1. **No Python tests currently gate CI.** Every "test-python: success" since this started is unverified.
2. The two `tests/test_evolution.py` failures previously triaged as "pre-existing on main, CI green, environment quirk" (`test_ga_evolve_basic`: fitness 5218.0 vs asserted range [-100, 100]; `test_parallel_with_scenario`) are **not exonerated by CI** — CI never ran them. They reproduce locally on a fresh Rust rebuild and may be genuine regressions (the fitness magnitude looks like a per-episode total where the test expects a per-step/normalized value — plausibly related to the RustFitnessEvaluator API drift noted in earlier sessions).

# Proposed fix (one PR, three parts)

1. **Workflow**: make the step fail-honest — `set -o pipefail` before the pipe (or write with `--junitxml`/redirect instead of piping to tee). Consider `--cov-fail-under` implications once real coverage numbers appear (adjust threshold with justification rather than deleting the gate).
2. **Collection errors**: make the suite collect cleanly without the `rl` extra — add `pytest.importorskip("torch")` (module-level) or equivalent guards to the 18 failing modules, consistent with the repo's existing skip conventions (`tests/conftest.py` has the requires_rust pattern to follow). Do NOT add torch to the CI test job without discussing cost — the guard approach matches the "fast tests" job intent.
3. **test_evolution failures**: diagnose and fix `test_ga_evolve_basic` + `test_parallel_with_scenario` — determine whether the fitness-scale change is a code regression (bisect against the RustFitnessEvaluator/reward changes) or a stale test expectation, and fix whichever is wrong. These will block the now-honest CI otherwise.

# Acceptance criteria

- [ ] A deliberately-failing test makes the `test-python` job fail (verified on the PR branch, then reverted).
- [ ] `pytest -m "not slow and not integration"` collects with 0 errors under a dev-only (no torch) env, runs to completion, and passes.
- [ ] test_evolution failures root-caused: fix committed with the diagnosis stated (regression vs stale expectation).
- [ ] Coverage gate behavior stated explicitly (passes at 75, or threshold adjusted with rationale).

# References

- Run 28740450491 (test-python job log, collection-error evidence)
- `.github/workflows/ci.yml` "Run fast tests" step
- tests/conftest.py requires_rust pattern (#189/#330)

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

The `test-python` CI job was **false-green**: `pytest ... | tee pytest-output.txt` without `pipefail` made the step's exit status tee's (always 0). Pytest was actually aborting at collection with 18 import errors and running **zero tests** — every "test-python: success" was unverified. This PR makes the job fail-honest, makes the fast suite collect and pass in the job's actual environment (dev-only, no torch), and root-causes every failure the false-green had been hiding. The red-run experiment then revealed **integration-test was also false-green** (pytest never even installed there); that is fixed too.

## Changes

**1. Workflow honesty (`.github/workflows/ci.yml`)**
- `set -o pipefail` in "Run fast tests" (tee kept for the step summary).
- Same fix in "Run integration tests", which had the identical pattern.
- integration-test job now installs `-e .[dev]`: pytest is a dev extra and was never installed there, so `uv run pytest` failed to spawn (exit 2) and tee swallowed it — **the pytest integration tests have never run in CI** (proven by the red run below, where integration-test failed with `error: Failed to spawn: pytest` the moment pipefail made it honest).
- `--cov-fail-under` lowered 75 → 45 with an in-file rationale: the dev-only job has no torch, so all RL-gated modules (`bucket_brigade/training/*`, etc.) measure 0%; **real coverage in this environment is 48%** (confirmed in CI: "Required test coverage of 45% reached. Total coverage: 48.00%"). The 75 gate was never actually enforced (coverage never ran). The gate is kept, set just below the measured baseline.

**2. Collection errors (18 → 0 torch-less)**
- Module-level `pytest.importorskip("torch")` guards added to **17 modules** that import torch directly or transitively, following the existing `tests/test_training.py` convention.
- `tests/test_code_generation.py` (the 18th) fails on `bucket_brigade.agents.archetypes_generated`, **not torch** — that module is produced by `scripts/generate_python.py` and is not committed. Guarded with `importorskip` on the generated module.

**3. test_evolution root cause (diagnosed, not just widened)**
- `test_ga_evolve_basic` (fitness 5218.0 vs asserted [-100, 100]): **stale test expectation, not a code regression.** Fitness is `GameResult.final_score`. It was redefined as the team payoff in 7f989e27 (Nov 4 2025, "±100 scale" — terminal one-shot value, when the bound was valid). Three days later, bucket-brigade-core commit **354219d9** ("refactor: Align Rust reward computation with Python per-step logic", an explicitly BREAKING, deliberate change) made ALL reward components (team, ownership, work/rest) accrue **per step**, so `final_score` now scales with episode length × num_agents: trivial_cooperation ≈ 13 nights × 4 agents × ~100 team reward ≈ 5218 — exactly the observed value. The per-step contract is what PPO training and the Python env rely on, so the code is correct; the test now derives its bound from the scenario parameters and documents the contract.
- `test_parallel_with_scenario`: **stale test.** `RustFitnessEvaluator` dropped the `scenario=` (Python object) kwarg for `scenario_name: str` + `num_agents` in 0e2fdb30 when evaluation moved to the Rust core. Test updated to the current API.

**4. Additional failures newly exposed by the honest suite (also root-caused)**
- `test_equilibrium.py::test_evaluate_symmetric_payoff_sequential`: `PayoffEvaluator.evaluate_symmetric_payoff` (sequential branch) returned `np.float32` — float32 reward arrays promote the accumulator — violating its declared `-> float` contract (`np.float32` is not an instance of `float`, unlike `np.float64`). The parallel branch already cast. Fixed with a `float()` cast in `bucket_brigade/equilibrium/payoff_evaluator.py`.
- `test_kstar_vs_trainability.py::test_committed_output_files_match_regeneration`: byte-identical JSON regeneration does not survive numpy 1.26 → 2.x (reduction changes shift the last 1–2 digits of pooled means and permutation p-values; verified the diff is last-ulp float noise only, and the committed **.md still matches byte-for-byte**). The JSON comparison is now a structural deep-compare with `rel=1e-9` on floats; keys, ints, strings, bools, and the .md remain exact.
- `tests/test_integration.py` (never run in CI, see above): the same two stale-test classes — 9 `RustFitnessEvaluator(scenario=...)` call sites updated to the `scenario_name` API, and pre-354219d9 fitness bounds replaced with a scenario-derived bound helper (roundtrip tolerance 1.5x → 2x, observed 1.52x between two small-sample Monte Carlo estimates).

## Red-run proof (acceptance criterion 1)

- Commit `9f535cf9` adds `tests/test_ci_red_proof_issue484.py` (deliberate `assert False`).
- **RED run: https://github.com/rjwalters/bucket-brigade/actions/runs/28741319226** — `test-python: failure`, log shows `FAILED tests/test_ci_red_proof_issue484.py::test_ci_goes_red_on_failure_issue484` with `1 failed, 836 passed, 29 skipped` (the job would previously have reported success here).
- Commit `113f54f8` reverts it. Both commits remain in history as the audit trail.
- Bonus finding: the same red run exposed `integration-test: failure` (`Failed to spawn: pytest`), proving that job was also false-green — fixed in `95836fed`.

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|-----------|--------|--------------|
| Deliberately-failing test makes test-python fail, then reverted | ✅ | Red run https://github.com/rjwalters/bucket-brigade/actions/runs/28741319226 on commit 9f535cf9; reverted in 113f54f8; both in history |
| Suite collects with 0 errors and passes under dev-only (no torch) env | ✅ | Fresh venv with `.[dev]` + Rust ext only: `842/878 tests collected (36 deselected)`, 0 errors; full run: **836 passed, 29 skipped, 0 failed** (locally and in the CI red run, modulo the deliberate failure) |
| test_evolution failures root-caused with diagnosis stated | ✅ | Stale expectations vs deliberate contract changes 354219d9 (per-step rewards) and 0e2fdb30 (scenario_name API) — see section 3 |
| Coverage gate behavior stated explicitly | ✅ | Measured 48% torch-less (locally and in CI); threshold set to 45 (not deleted) with in-file comment — see section 1 |

## Test Plan

- Reproduced the exact 18 CI collection errors in a fresh torch-less venv, then verified 0 errors after the guards.
- `tests/test_evolution.py`: 33/33 pass torch-less.
- Full fast suite torch-less: 836 passed, 29 skipped, coverage 48% (gate 45 passes) — matched by the CI red run's numbers.
- `pytest --run-integration tests/test_integration.py` torch-less: 7 passed, 3 skipped (slow-marked), 0 failed.
- Guarded modules verified to still run with torch present (39/39 spot-check).
- `ruff format` + `ruff check` clean on all touched files.

Closes #484

### CODE REVIEW

**Comment by rjwalters** [comment]
✅ **Approved!**

This is the most safety-critical PR of the week and it holds up under independent verification. Every load-bearing claim was re-derived, not taken on trust.

## Independent verification of the stale-test diagnosis (highest-stakes item)

- **Commit 354219d9 exists and says exactly what the PR claims**: "refactor: Align Rust reward computation with Python per-step logic" with an explicit "BREAKING CHANGE: Rust now computes all rewards per-step (not end-of-game)" body. This was a deliberate contract change, not an accidental regression — the fitness change is correct behavior and the old `[-100, 100]` assertion was stale.
- **The 5218 arithmetic checks out against actual scenario params**: `core.SCENARIOS["trivial_cooperation"]` has `team_reward_house_survives = 100.0`, `min_nights = 12` (≈13 steps), 4 agents → ~5200 from team reward alone; the +18 residual is consistent with per-step rest/work (±0.5) and ownership terms. 5218 is exactly what the per-step contract predicts.
- **The new bound is genuinely scenario-derived**: team/ownership/work-rest terms all read from the scenario object; the only literal is `max_steps = 100`, which I verified matches the actual safety cap at `bucket_brigade/evolution/fitness_rust.py:84`. Resulting bound (~48,200) is loose but that is appropriate for a sanity assertion.
- **API diagnosis confirmed**: commit 0e2fdb30 shows `RustFitnessEvaluator.__init__(scenario_name: str = ..., num_agents: int = ...)` — the `scenario=` kwarg was dropped there, so the 9 `test_integration.py` call sites and `test_parallel_with_scenario` were stale, as claimed.

## Red-run proof verified

- Run 28741319226 is on commit 9f535cf9 (the deliberate-failure commit): **test-python = failure**, log shows `FAILED tests/test_ci_red_proof_issue484.py::test_ci_goes_red_on_failure_issue484` and `= 1 failed, 836 passed, 29 skipped, 36 deselected =`. Exactly as claimed.
- Revert commit 113f54f8 is in the PR history; the proof file is absent at head.
- Bonus finding confirmed in the same red-run logs: **integration-test = failure** with `error: Failed to spawn: pytest ... exit code 2` — hard proof that job was also false-green (pytest never installed without `[dev]`).

## Integration-test now genuinely runs

Green run 28741536276 (on PR head 95836fed) integration-test log: `collected 10 items ... 7 passed, 3 skipped in 3.16s`. First time these tests have actually run in CI.

## Local reproduction (torch-less dev-only env, matching CI)

- Collection: **842/878 collected, 36 deselected, 0 errors** — matches the claim exactly.
- `tests/test_evolution.py` + `tests/test_equilibrium.py` + `tests/test_kstar_vs_trainability.py`: 57 passed, 2 skipped.
- kstar regeneration test passes — committed artifacts still regenerate consistently under the numeric comparison; the .md remains byte-exact.
- `pytest --run-integration tests/test_integration.py`: 7 passed, 3 skipped.
- `ruff check` + `ruff format --check` clean on all 23 touched files (the 42 repo-wide ruff findings are pre-existing outside CI's lint scope, not introduced here).
- Guard convention verified: module-level `torch = pytest.importorskip("torch")` before the failing import, matching the existing `tests/test_training.py:9` pattern; `tests/test_code_generation.py` correctly guards `archetypes_generated` (a codegen artifact), not torch.
- pipefail fix is genuinely required: ci.yml uses the default shell (`bash -e {0}`), which does NOT set pipefail — confirmed empirically by the red run.

## Notes (non-blocking)

- **Coverage gate 45 vs measured 48**: the 3-point headroom means small coverage regressions (up to 3 pts) pass silently. Acceptable — the gate's purpose here is catching gross regressions (e.g., collection silently dying again), and 45 will bind hard on that. Worth revisiting upward if torch-less coverage grows.
- The roundtrip tolerance widening (1.5x → 2x) is justified with observed data (1.52x) and the assertion retains its intent (`fitness > 0` plus the ratio cap).

## CI / merge state

All 6 checks green on head 95836fed (test-python, integration-test, lint-python, pre-commit, test-web, build-web). Merge state: CLEAN.

This PR restores the meaning of "CI green" for the repo, with a complete audit trail. Excellent work.
